model/src/gateway/event/gateway.rs

```diff
@@ -28,9 +30,122 @@ enum Field {
     T,
 }
 
-struct GatewayEventVisitor;
+/// A deserializer that deserializes into a `GatewayEvent` by cloning some bits
+/// of scanned information before the actual deserialisation.
+///
+/// This is the owned version of [`GatewayEventDeserializer`].
+///
+/// You should use this if you're using a mutable deserialization library
+/// like `simd-json`.
+///
+/// [`GatewayEventDeserializer`]: struct.GatewayEventDeserializer.html
+pub struct GatewayEventDeserializerOwned {
+    event_type: Option<String>,
+    op: u8,
+}
+
+impl GatewayEventDeserializerOwned {
+    pub fn from_json(input: &str) -> Option<Self> {
+        let deser = GatewayEventDeserializer::from_json(input)?;
+        let GatewayEventDeserializer { event_type, op } = deser;
+
+        Some(Self {
+            event_type: event_type.map(ToOwned::to_owned),
+            op,
+        })
+    }
+}
+
+/// A deserializer that deserializes into a `GatewayEvent` by borrowing some bits
+/// of scanned information before the actual deserialisation.
+///
+/// This is the borrowed version of [`GatewayEventDeserializerOwned`].
+///
+/// You should use this if you're using an immutable deserialization library
+/// like `serde_json`.
+///
+/// [`GatewayEventDeserializerOwned`]: struct.GatewayEventDeserializerOwned.html
+pub struct GatewayEventDeserializer<'a> {
+    event_type: Option<&'a str>,
+    op: u8,
+}
 
-impl<'de> Visitor<'de> for GatewayEventVisitor {
+impl<'a> GatewayEventDeserializer<'a> {
+    // Create a gateway event deserializer with some information found by
+    // scanning the JSON payload to deserialise.
+    //
+    // This will scan the payload for the opcode and, optionally, event type if
+    // provided. The opcode key ("op"), must be in the payload while the event
+    // type key ("t") is optional and only required for event ops.
+    pub fn from_json(input: &'a str) -> Option<Self> {
+        let op = Self::find_opcode(input)?;
+        let event_type = Self::find_event_type(input);
+
+        Some(Self { event_type, op })
+    }
+
+    fn find_event_type(input: &'a str) -> Option<&'a str> {
+        // Discord seems to usually (always?) provide the type at the end of the
+        // payload, so let's search from the right...
+        let from = input.rfind(r#""t":"#)? + 4;
```

Comment:
Could this lead to erronous deserilazation given that the payload is not the final element and eg a user message contains that text?

Reply:
The quotation marks would be escaped in that scenario